src/domain/flow-system/src/entities/shared/batching_rule.rs

```diff
@@ -0,0 +1,134 @@
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use chrono::{DateTime, Duration, Utc};
+use kamu_core::{DatasetChangesService, InternalError, ResultIntoInternal};
+use thiserror::Error;
+
+use crate::{FlowResult, FlowTrigger};
+
+/////////////////////////////////////////////////////////////////////////////////////////
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct BatchingRule {
+    min_records_to_await: u64,
+    max_batching_interval: Duration,
+}
+
+impl BatchingRule {
+    const MAX_BATCHING_INTERVAL_HOURS: i64 = 24;
+
+    pub fn new_checked(
+        min_records_to_await: u64,
+        max_batching_interval: Duration,
+    ) -> Result<Self, BatchingRuleValidationError> {
+        if min_records_to_await == 0 {
+            return Err(BatchingRuleValidationError::MinRecordsToAwaitNotPositive);
+        }
+
+        let max_possible_interval = Duration::hours(Self::MAX_BATCHING_INTERVAL_HOURS);
+        if max_batching_interval > max_possible_interval {
+            return Err(BatchingRuleValidationError::MaxIntervalAboveLimit);
+        }
+
+        Ok(Self {
+            min_records_to_await,
+            max_batching_interval,
+        })
+    }
+
+    #[inline]
+    pub fn min_records_to_await(&self) -> u64 {
+        self.min_records_to_await
+    }
+
+    #[inline]
+    pub fn max_batching_interval(&self) -> &Duration {
+        &self.max_batching_interval
+    }
+
+    pub async fn evaluate(
+        &self,
+        flow_start_time: DateTime<Utc>,
+        evaluation_time: DateTime<Utc>,
+        triggers: &[FlowTrigger],
+        dataset_changes_service: &dyn DatasetChangesService,
+    ) -> Result<BatchingRuleEvaluation, InternalError> {
+        // TODO: it's likely assumed the accumulation is per each input separately, but
+        // for now count overall number
+        let mut accumulated_records_count = 0;
+        let mut watermark_modified = false;
+
+        // Scan each accumulated trigger to decide
+        for trigger in triggers {
+            if let FlowTrigger::InputDatasetFlow(trigger) = trigger {
+                match &trigger.flow_result {
+                    FlowResult::Empty => {}
+                    FlowResult::DatasetUpdate(update) => {
+                        // Compute increment since the first trigger by this dataset.
+                        // Note: there might have been multiple updates since that time.
+                        // We are only recording the first trigger of particular dataset.
+                        let increment = dataset_changes_service
+                            .get_increment_since(&trigger.dataset_id, update.old_head.as_ref())
+                            .await
+                            .int_err()?;
```

Comment:
If we are still allowing multiple `InputDatasetFlow` triggers per input - I don't see what stopping this loop from double / tripple-counting the same input.

Reply:
We are storing only 1 trigger per dataset